### PR TITLE
python-curl: Add variant for Python3

### PR DESCRIPTION
**Pull request by BKPepe**
Maintainer: @valdi74 
Compiled tested: 

- Xiaomi Mi WiFi R3G, ramips, mt7621, OpenWrt 18.06.02
- Turris Omnia, mvebu (cortex-a9_vfpv3) , OpenWrt SNAPSHOT, r9420-c17a68c

Run tested: 

- Xiaomi Mi WiFi R3G, ramips, mt7621, OpenWrt 18.06.02
- Turris Omnia, mvebu (cortex-a9_vfpv3) , OpenWrt SNAPSHOT, r9420-c17a68c

Description:
- add Python3 variant (which was suggested in PR https://github.com/openwrt/packages/pull/4478#discussion_r138832948)
- simplify Makefile

### CODE REVIEW

**Comment by BKPepe** [comment]
I don't know, how it is possible that PR https://github.com/openwrt/packages/pull/6191 was merged as you can see from my 2nd commit, I have applied it as well, but it is just wrong **if you didn't compile libcurl yourself and install it!** Since 30th December 2016 is used mbedTLS as default for curl https://github.com/openwrt/openwrt/commit/558680012dd79de0289aeeb74a72163916cd9f33

```
root@OpenWrt:~# opkg install python-curl
Installing python-curl (7.43.0.2-1) to root...
Downloading http://downloads.openwrt.org/snapshots/packages/arm_cortex-a9_vfpv3/packages/python-curl_7.43.0.2-1_arm_cortex-a9_vfpv3.ipk
Installing libcurl4 (7.64.0-1) to root...
Downloading http://downloads.openwrt.org/snapshots/packages/arm_cortex-a9_vfpv3/base/libcurl4_7.64.0-1_arm_cortex-a9_vfpv3.ipk
Configuring libcurl4.
Configuring python-curl.
root@OpenWrt:~# python
Python 2.7.15 (default, Mar 02 2019, 09:09:01) 
[GCC 7.4.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import curl
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/site-packages/curl/__init__.py", line 7, in <module>
    import sys, pycurl
ImportError: pycurl: libcurl link-time ssl backend (none/other) is different from compile-time ssl backend (openssl)
>>> 
```

To get it work, you need to remove libcurl4 and use the libcurl, which was compiled together with python-curl! I doubt that many users did it.

The thing is that if I just add [these](https://github.com/pycurl/pycurl/commit/3a0ab495e82cc1432c701292fff2e951234d0143) patches to detect it, then you can see in log from [circleci](https://circleci.com/gh/openwrt/packages/1864).

```
Using curl-config (libcurl 7.64.0)
Not using an SSL library

```

Nowadays SSL is must have. Otherwise python3-curl works:
screenshot from 2019-03-04 15-31-53

**Comment by BKPepe** [comment]
I don't know, how it is possible that PR https://github.com/openwrt/packages/pull/6191 was merged as you can see from my 2nd commit, I have applied it as well, but it is just wrong **if you didn't compile libcurl yourself and install it!** Since 30th December 2016 is used mbedTLS as default for curl https://github.com/openwrt/openwrt/commit/558680012dd79de0289aeeb74a72163916cd9f33

```
root@OpenWrt:~# opkg install python-curl
Installing python-curl (7.43.0.2-1) to root...
Downloading http://downloads.openwrt.org/snapshots/packages/arm_cortex-a9_vfpv3/packages/python-curl_7.43.0.2-1_arm_cortex-a9_vfpv3.ipk
Installing libcurl4 (7.64.0-1) to root...
Downloading http://downloads.openwrt.org/snapshots/packages/arm_cortex-a9_vfpv3/base/libcurl4_7.64.0-1_arm_cortex-a9_vfpv3.ipk
Configuring libcurl4.
Configuring python-curl.
root@OpenWrt:~# python
Python 2.7.15 (default, Mar 02 2019, 09:09:01) 
[GCC 7.4.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import curl
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/site-packages/curl/__init__.py", line 7, in <module>
    import sys, pycurl
ImportError: pycurl: libcurl link-time ssl backend (none/other) is different from compile-time ssl backend (openssl)
>>> 
```

If you want to get it to work, you need to remove libcurl4 (OpenWrt base packages) and use the libcurl, which was compiled together with python-curl! I doubt that many users did it.

The thing is that if I just add [these](https://github.com/pycurl/pycurl/commit/3a0ab495e82cc1432c701292fff2e951234d0143) patches to detect it, then you can see in log from [circleci](https://circleci.com/gh/openwrt/packages/1864).

```
Using curl-config (libcurl 7.64.0)
Not using an SSL library

```

Nowadays SSL is must have. Otherwise, python3-curl works:
screenshot from 2019-03-04 15-31-53

I'll look, what's necessary to be done to compile python-curl as well with mbedTLS.

**Comment by BKPepe** [comment]
It's over 2 weeks and my PR is not reviewed in pycurl for adding mbedTLS. I was able to compile it and test it on mipsel_24kc, ramips - mt7621, Xiaomi Mi WiFi R3G, OpenWrt 18.06.02, so this PR is ready for review.

**Comment by jefferyto** [comment]
A few comments:

* If you set
  ```
  PYTHON_PKG_SETUP_ARGS:=--with-mbedtls
  PYTHON3_PKG_SETUP_ARGS:=--with-mbedtls
  ```
  then you shouldn't need to override `PyBuild/Compile` / `Py3Build/Compile`.

* I would prefer to see upstream accept the mbedTLS patch before we carry it, but I don't know how others feel about that.

* Would it make sense to add both mbedTLS and OpenSSL variants of this package (and have them `DEPENDS` the correct library), and maybe also a nossl variant? Off the top of my head I can think of OpenVPN (`openvpn-openssl` / `openvpn-mbedtls` / `openvpn-nossl`), but I'm pretty sure there are other packages that do something similar.

Also ping @commodo :wink:

**Comment by BKPepe** [comment]
> PYTHON_PKG_SETUP_ARGS:=--with-mbedtls
> PYTHON3_PKG_SETUP_ARGS:=--with-mbedtls

Good point. Thanks!

> I would prefer to see upstream accept the mbedTLS patch before we carry it, but I don't know how others feel about that.

Me too. :(

File: lang/python/python-curl/Makefile [hunk 1]
```diff
@@ -6,40 +6,61 @@ include $(TOPDIR)/rules.mk
...
 
-PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
-PKG_SOURCE_URL:=https://dl.bintray.com/pycurl/pycurl/
+PKG_SOURCE:=pycurl-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/p/pycurl
 PKG_HASH:=0f0cdfc7a92d4f2a5c44226162434e34f7d6967d3af416a6f1448649c09a25a4
 
+PKG_LICENSE:=LGPL-2.1
+PKG_LICENSE_FILE:=COPYING-LGPL
+PKG_MAINTAINER:=Waldemar Konik <informatyk74@interia.pl>
```
Comment (neheb) [inline lang/python/python-curl/Makefile]: Above PKG_LICENSE please.

Comment (BKPepe) [reply lang/python/python-curl/Makefile]: Right, changed.

**Comment by BKPepe** [comment]
@valdi74 Any objections?

**Comment by valdi74** [comment]
> @valdi74 Any objections?

No objections, OK for me. I have no rights to merge pull requests.

**Comment by jefferyto** [comment]
This may be a silly question, but I'll ask anyway: Does pycurl have a test suite or something we can run to verify the mbedTLS changes have no issues? I understand you have tested it, but security is tricky and I wouldn't want any security flaws to be introduced accidentally.

Also, I would still prefer to have a choice between using mbedTLS and OpenSSL. I know mbedTLS is the "default", but using myself as an example, I only use OpenSSL on my devices (using luci-ssl-openssl, openvpn-openssl, etc.) and I would not want to have to install mbedTLS just for one package.

**Comment by BKPepe** [comment]
Good question. I agree about security, but I will try to reach people from pycurl to review the pull request.

The second paragraph applies as well for Turris routers. As we use by default OpenSSL instead of mbedTLS, but simply patch which will change mbedtls to openssl would be just fine for us, but I'm working on it. I like you suggestion and hopefulyl it will not be a problem at all.

**Comment by jefferyto** [comment]
@BKPepe thanks!

**Pull request merged**